### PR TITLE
Plugins: Remove support for deprecated backend plugin protocol version

### LINKED ISSUE #33959: Backend plugins: Remove support for deprecated backend plugin protocol (v1)
**What would you like to be added**:
Remove support for deprecated backend plugin protocol (v1) including all code referencing github.com/grafana/grafana-plugin-model.

**Why is this needed**:
We did archive https://github.com/grafana/grafana-plugin-model and deprecated [plugin protocol version 1](https://github.com/grafana/grafana/blob/a015bb3a23d13a13429014a6efbdfc4678e04f11/pkg/plugins/backendplugin/grpcplugin/client_v1.go#L22-L23) for Grafana v7 and now it's time to remove it completely. 

For example the following files needs to be deleted/updated:
- https://github.com/grafana/grafana/blob/main/pkg/plugins/backendplugin/grpcplugin/client_v1.go
- https://github.com/grafana/grafana/blob/main/pkg/plugins/backendplugin/grpcplugin/grpc_plugin.go#L66-L71
- https://github.com/grafana/grafana/blob/main/pkg/plugins/backendplugin/grpcplugin/client.go#L58

### PR DESCRIPTION
**Pull request by idafurjes**
**What this PR does / why we need it**:
Remove support for deprecated backend plugin protocol (v1)

**Which issue(s) this PR fixes**:

Fixes #33959

**Special notes for your reviewer**:

# Release notice breaking change
Removes support for deprecated backend plugin protocol (v1) including usage of github.com/grafana/grafana-plugin-model.

### CODE REVIEW

File: pkg/plugins/backendplugin/grpcplugin/grpc_plugin.go [hunk 1]
```diff
@@ -58,16 +58,9 @@ func (p *grpcPlugin) Start(ctx context.Context) error {
...
-		p.pluginClient, err = newClientV2(p.descriptor, p.logger, rpcClient)
-		if err != nil {
-			return err
-		}
-	} else {
-		p.pluginClient, err = newClientV1(p.descriptor, p.logger, rpcClient)
-		if err != nil {
-			return err
-		}
+	p.pluginClient, err = newClientV2(p.descriptor, p.logger, rpcClient)
```
Comment (marefr) [inline pkg/plugins/backendplugin/grpcplugin/grpc_plugin.go:61]: I think an error returned if `p.client.NegotiatedVersion() < 2` would be nice. Something along the lines "Plugin uses a version of Grafana's backend plugin system that is no longer supported"

File: pkg/plugins/datasource_plugin.go [hunk 2]
```diff
@@ -73,13 +72,6 @@ func (p *DataSourcePlugin) DataQuery(ctx context.Context, dsInfo *models.DataSou
 	return endpoint.Query(ctx, dsInfo, query)
 }
 
-func (p *DataSourcePlugin) onLegacyPluginStart(pluginID string, client *grpcplugin.LegacyClient, logger log.Logger) error {
-	p.legacyClient = client
-	p.logger = logger
```
Comment (marefr) [inline pkg/plugins/datasource_plugin.go:78]: Are these struct fields still in use or can we remove them as well?

File: pkg/plugins/backendplugin/grpcplugin/client.go [hunk 3]
```diff
@@ -55,8 +55,7 @@ type StartFunc func(pluginID string, client *Client, logger log.Logger) error
 
 // PluginStartFuncs functions called for plugin when started.
 type PluginStartFuncs struct {
-	OnLegacyStart LegacyStartFunc
-	OnStart       StartFunc
+	OnStart StartFunc
```
Comment (marefr) [inline pkg/plugins/backendplugin/grpcplugin/client.go:58]: `LegacyClient` below can be removed as well.

**Review by marefr — CHANGES_REQUESTED** [review]
Looks good in general :+1: Added some comments. 

This this file and its tests can be removed as well: https://github.com/grafana/grafana/blob/main/pkg/plugins/datasource_plugin_wrapper.go

File: pkg/plugins/datasource_plugin.go (hunk 2 above)
Comment (idafurjes) [reply pkg/plugins/datasource_plugin.go:78]: I think they are

**Comment by marefr** [comment]
@idafurjes please change from draft to ready for review when this is ready for review - you can re-request review from me as well 
image

Thanks

File: pkg/plugins/backendplugin/grpcplugin/client.go [hunk 4]
```diff
@@ -47,16 +47,12 @@ func newClientConfig(executablePath string, env []string, logger log.Logger,
...
-type LegacyStartFunc func(pluginID string, client *LegacyClient, logger log.Logger) error
-
 // StartFunc callback function called when a plugin with current plugin protocol version is started.
 type StartFunc func(pluginID string, client *Client, logger log.Logger) error
 
 // PluginStartFuncs functions called for plugin when started.
 type PluginStartFuncs struct {
-	OnLegacyStart LegacyStartFunc
-	OnStart       StartFunc
+	OnStart StartFunc
```
Comment (marefr) [inline pkg/plugins/backendplugin/grpcplugin/client.go:55]: After consideration I think we should remove everything related/referencing `DefaultProtocolVersion` in this file and change `handshake.ProtocolVersion = grpcplugin.ProtocolVersion`. With this we should get rid of the last reference to github.com/grafana/grafana-plugin-model and we can tidy up the go.mod file to get rid of the github.com/grafana/grafana-plugin-model dependency altogether.

**Review by marefr — CHANGES_REQUESTED** [review]
Looks great. I have a last request of changes, see comment.

File: pkg/plugins/backendplugin/grpcplugin/client.go (hunk 4 above)
Comment (idafurjes) [reply pkg/plugins/backendplugin/grpcplugin/client.go:46]: I removed the DefaultProtocolVersion as well from renderer and backend plugin

**Review by marefr — APPROVED** [review]
LGTM great :+1: 

We should label this PR with "breaking change" so that it will show up in the changelog/release notes as a breaking change.

PR title could probably be improved given that it will be included as a breaking change. Something like "Plugins: Remove support for deprecated backend plugin protocol version"

**Pull request merged**